circuits/src/zk_circuits/valid_wallet_update.rs

```diff
@@ -0,0 +1,1242 @@
+        ));
+    }
+
+    // ----------------------
+    // | Deposit Test Cases |
+    // ----------------------
+
+    /// Tests a valid deposit into the wallet that adds a new balance
+    #[test]
+    fn test_external_transfer__valid_deposit_new_balance() {
+        let mut old_wallet = INITIAL_WALLET.clone();
+        let new_wallet = INITIAL_WALLET.clone();
+
+        // Remove the first balance from the old wallet
+        old_wallet.balances[0] = Balance::default();
+
+        // Transfer a brand new mint into the new wallet
+        let deposit_mint = new_wallet.balances[0].mint.clone();
+        let deposit_amount = new_wallet.balances[0].amount;
+
+        let transfer = ExternalTransfer {
+            mint: deposit_mint,
+            amount: BigUint::from(deposit_amount),
+            direction: ExternalTransferDirection::Deposit,
+            account_addr: BigUint::from(0u8),
+        };
+
+        assert!(constraints_satisfied_on_wallets(
+            old_wallet, new_wallet, transfer
+        ));
+    }
+
+    /// Tests a valid deposit into the wallet that adds to an existing balance
+    #[test]
+    fn test_external_transfer__valid_deposit_existing_balance() {
+        let old_wallet = INITIAL_WALLET.clone();
+        let mut new_wallet = INITIAL_WALLET.clone();
+
+        // Transfer units of an existing mint into the wallet
+        let deposit_mint = new_wallet.balances[1].mint.clone();
+        let deposit_amount = new_wallet.balances[1].amount;
+
+        new_wallet.balances[1].amount += deposit_amount;
+
+        let transfer = ExternalTransfer {
+            mint: deposit_mint,
+            amount: BigUint::from(deposit_amount),
+            direction: ExternalTransferDirection::Deposit,
+            account_addr: BigUint::from(0u8),
+        };
+
+        assert!(constraints_satisfied_on_wallets(
+            old_wallet, new_wallet, transfer
+        ));
+    }
+
+    /// Tests an invalid deposit in which the balance is updated incorrectly
+    #[test]
+    fn test_external_transfer__invalid_deposit_balance_update() {
+        let old_wallet = INITIAL_WALLET.clone();
+        let mut new_wallet = INITIAL_WALLET.clone();
+
+        // Transfer units of an existing mint into the wallet
+        let deposit_mint = new_wallet.balances[1].mint.clone();
+        let deposit_amount = new_wallet.balances[1].amount;
+
+        // Prover adds one more unit than the transfer is for
+        new_wallet.balances[1].amount += deposit_amount + 1;
+
+        let transfer = ExternalTransfer {
+            mint: deposit_mint,
+            amount: BigUint::from(deposit_amount),
+            direction: ExternalTransferDirection::Deposit,
+            account_addr: BigUint::from(0u8),
+        };
+
+        assert!(!constraints_satisfied_on_wallets(
+            old_wallet, new_wallet, transfer
+        ));
+    }
+
+    /// Tests an invalid deposit in which the prover adds an unrelated balance
+    /// as well as the updated balance
```

Comment:
is it worth doing a withdrawal test case like this as well, i.e. "withdraw from one balance but also update another"?

Reply:
added